### PR TITLE
chore: inverse unit tests for CalcInAmtGivenOut and CalcOutAmtGivenIn with zero swap fee

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## Description

Trying to investigate https://github.com/osmosis-labs/osmosis/issues/1196 by incrementally adding more unit tests

These tests do not allow us to conclude anything with regards to #1196 yet. The tests added so far have a swap fee of 0. I'm planning to continue adding unit tests with a non-zero swap fee. Also, test other functionality and methods that are being called in `ExitSwapShareAmountIn` and `JoinSwapExactAmountIn`

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Review by alexanderbez — COMMENTED** [review]
So what did we find out here? There's a test case commented out in `TestJoinSwapExactAmountInConsistency`.

**Comment by p0mvn** [comment]
> So what did we find out here? There's a test case commented out in `TestJoinSwapExactAmountInConsistency`.

Nothing yet, I'm planning to continue increasing tests / coverage for all methods called within `JoinSwapExactAmountIn`.

Currently, it's hard to determine where exactly the error is happening because we are missing unit tests for each of the methods called by`JoinSwapExactAmountIn`. By increasing the number of unit tests for each of the called methods, we should make it more likely to find the problem

File: x/gamm/pool-models/balancer/util_test.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+func createTestContext(t *testing.T) sdk.Context {
+	db := dbm.NewMemDB()
+	logger := log.NewNopLogger()
+
+	ms := rootmulti.NewStore(db, logger)
+
+	return sdk.NewContext(ms, tmtypes.Header{}, false, logger)
+}
```
Comment (mattverse) [inline x/gamm/pool-models/balancer/util_test.go:42]: nit: we can also add these to app/app_testing/test_suite.go!

**Review by mattverse — COMMENTED** [review]
Hrmm interesting. So I think these test cases tell us that running `CalcInAmtGivenOut` and `CalcOutAmtGiven` straight away returns the correct value we expect, and that the problem is coming from other steps?

**Comment by p0mvn** [comment]
> Hrmm interesting. So I think these test cases tell us that running `CalcInAmtGivenOut` and `CalcOutAmtGiven` straight away returns the correct value we expect, and that the problem is coming from other steps?

It's unclear yet. I have tested with 0 swap fees so far. I'm planning to add the tests with a non-zero swap fee in a separate PR

**Comment by mattverse** [comment]
Sounds great! Can we address that this is a test for zero fee either in the comment or in the test case name?

**Comment by alexanderbez** [comment]
Can we leave this PR in Draft mode to indicate that it's still being worked on?

File: x/gamm/pool-models/balancer/util_test.go (hunk 1 above)
Comment (p0mvn) [reply x/gamm/pool-models/balancer/util_test.go:42]: The files in `test_suite.go` seem to belong to the keeper test suite.

On the other hand, these are low-level balancer pool test utilities. I briefly looked at the `test_uite.go`, and I don't think there are any applications for these utilities there. So I think we might want to keep them here. Let me know if you agree

**Comment by p0mvn** [comment]
> Sounds great! Can we address that this is a test for zero fee either in the comment or in the test case name?

Done! I also created an issue to continue with the tests with a non-zero swap fee: https://github.com/osmosis-labs/osmosis/issues/1288

@alexanderbez this is ready for review. I'm trying to work on this in small parts / PRs to make it easier to review

File: x/gamm/pool-models/balancer/util_test.go (hunk 1 above)
Comment (mattverse) [reply x/gamm/pool-models/balancer/util_test.go:42]: oh yeah agree! Forgot that we're not using test suite while I was writing that suggestion!

**Review by alexanderbez — APPROVED** [review]
🌮

**Review by ValarDragon — APPROVED** [review]

**Pull request merged**